File: 0383-ransom-note/0383-ransom-note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> map;
        for (char ch : magazine) {
            map[ch]++;
        }

        for (char ch : ransomNote) {
            if (map.find(ch) == map.end()) {
                return false;
            } else if (map[ch] == 0) {
                return false;
            } else {
                map[ch]--;
            }
        }
        return true;
    }
};

File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        '''
        turn magazine into a dictionary.
        takes each letter from the dictionary.
        '''

        # 1. turn magazine into a dictionary
        map = Counter(magazine)

        # 2. try to take the item from dictionary
        for ch in ransomNote:
            # 3. if item not in magazine, then ransomNote can't be build with that magazine
            if ch not in map:
                return False
            # 4. if item already empty, then ransomNote can't be build with that magazine
            elif map[ch] == 0:
                return False
            # 5. if item exists and not empty, take that item
            else:
                map[ch] -= 1
        
        # 6. if ransomNote can be builded (iteration finish), return True
        return True